src/torchcodec/decoders/_core/VideoDecoder.cpp

```diff
@@ -391,7 +388,7 @@ torch::Tensor VideoDecoder::getKeyFrameIndices() {
 void VideoDecoder::addStream(
     int streamIndex,
     AVMediaType mediaType,
-    const torch::Device& device,
+    DeviceInterface* device,
```

Comment:
I think it's easier if `addStream()` still accepts a `const torch::Device&`. That means that `addStream()` would be responsible for calling `createDeviceInterface()`, which I think is fine. That would keep our concept of devices as an implementation detail for `VideoDecoder`. From the outside, we accept a `torch::Device` and internally we create the corresponding special logic for decoding on that device.

Reply:
done. Note that this required few other changes for `DeviceInterface`. For example, I switch from `std::string` to `torch::DeviceType` in device interface registration. This also required to introduce a new `createTorchDevice` to be able to filter out unsupported devices against registered device interfaces early.